mips: Import of mips32r2_common module

neiljay opened the pull request:
Add support for the mips32r2 architecture to RIOT.

cpu/mips32r2_common adds base architecture support for mips32r2 cores it can be
built in it own right as a 'CPU', but is dependant on a bootloader (like u-boot)
to have bootstrapped the system, this has been tested on a 'malta' FPGA system
(BOARD=mips-malta) with various mips32r2 compliant cores (interAptiv, P5600,
etc).

NOTE: This port requires the use of the GCC based MIPS Codescape SDK toolchain
available here:

https://community.imgtec.com/developers/mips/tools/codescape-mips-sdk/download-codescape-mips-sdk-essentials/

This has been tested against the following examples:
hello-world
default
ipc_pingpong
timer_periodic_wakeup
riot_and_cpp

---

Note this is a re-work of #5885 with review comments addressed but now broken down into 3 separate PR's of which this is the first.

miri64 commented:
Took the freedom to rename the PR :-)

neiljay commented:
@miri64 No worries, much nicer ;-)

haukepetersen commented on cpu/mips32r2_common/include/cpu.h:
```diff
@@ -0,0 +1,61 @@
+ */
+static inline void cpu_init_early(void)
+{
+    /* This function must exist else RIOT won't compile */
```
Is this function mandatory by some parts of the MIPS toolchain or similar? RIOT itself does not care about it...

neiljay replied:
Nope, looks like that one can go.

haukepetersen commented on boards/mips-malta/Makefile.include:
```diff
@@ -0,0 +1,3 @@
+export CPU = mips32r2_common
+export INCLUDES += -I$(RIOTBOARD)/$(BOARD)/include/
+export CFLAGS += -mhard-float -mdsp
```
How about a simple variable like `USE_HARD_FLOAT` and then checking that in the `mips_common/Makefile.include` and set the CFLAGS there?

haukepetersen commented on cpu/mips32r2_common/thread_arch.c:
```diff
@@ -0,0 +1,399 @@
+    p--;
+    *p-- = STACK_END_PAINT;
+
+
```
newline...

neiljay replied:
?? where ? Do you want the double newline removing #74 or one adding between #71 and #72 ?

haukepetersen replied:
sorry, I meant to remove the double newline l74

haukepetersen commented:
Sorry for the late review and about being so picky about formatting and such... 

IMHO the PR looks quite good to me, mainly minor things that I noticed. 

Now when trying to compile it, I get and error resulting from `MIPS_ELF_ROOT` not being set on my system. I simply downloaded the `mips-mti-elf-` `MTI Bare Metal` binaries tarball and used it right away. Seems like there is some more configuration effort needed here?!

neiljay commented:
@haukepetersen 

Thanks for the review.

>Now when trying to compile it, I get and error resulting from MIPS_ELF_ROOT not being set on my system. 

There are two toolchains for MIPS pre r6 (mips-mti-elf) and r6 onwards (mips-img-elf), MIPS_ELF_ROOT thus needs setting appropriately, I did add default paths to the makefile but Oleg asked me to remove them:

`cpu/mips_pic32mx/Makefile`
`@@ -0,0 +1,15 @@ `
`+ifeq ("$(OS)","Windows_NT")`
`+MIPS_ELF_ROOT?=C:\cross-mips-mti-elf`
`+else`
`+MIPS_ELF_ROOT?=/opt/imgtec/Toolchains/mips-mti-elf/2016.05-03`
`+endif`

@OlegHahm on Sep 29 Member

I don't think we want to have absolute paths here. :wink:

@neiljay on Sep 30

They are '?=' ie defaults if nothing else is set and are the default locations the toolchain installs too, but I can remove them if that's what you prefer.

@OlegHahm on Oct 13 Member

Yes, I would prefer to remove the absolute paths.

neiljay commented on boards/mips-malta/Makefile.features:
```diff
@@ -0,0 +1,3 @@
+FEATURES_PROVIDED += periph_timer
+FEATURES_PROVIDED += periph_uart
+FEATURES_PROVIDED += cpp
```
Will do.

OlegHahm commented:
Hm, okay, maybe I misjudged. Not sure. On the one hand, I still don't like to have these absolute paths in the build system, on the other hand that's probably the most convenient to do for most of the users.

@haukepetersen, do you have a preference?

neiljay commented on boards/mips-malta/include/board.h:
```diff
@@ -0,0 +1,46 @@
+ * @brief Set how many increments of the count register per uS
+ *        needed by timer code
+ */
+#define TICKS_PER_US (15)
```
The timer uses it, its the number of CPU core ticks per uS, so its not necessarily peripheral specific, peripherals could be (and often are) running from a different clock.

neiljay commented on boards/mips-malta/include/periph_conf.h:
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2016, Imagination Technologies Limited and/or its
+ *                 affiliated group companies.
+ * This file is subject to the terms and conditions of the GNU Lesser
```
ok.

neiljay commented on boards/mips-malta/malta.c:
```diff
@@ -0,0 +1,28 @@
+#define MIPS_MALTA_ADDR    (0xbf000500)
+#define MIPS_MALTA_VAL_RST (0x42)
+
+static void malta_reset(void)
```
No, they are specific to the Malta FPGA platform only.

haukepetersen replied:
ah ok.

neiljay commented on boards/mips-malta/include/board.h:
```diff
@@ -0,0 +1,46 @@
+ * @file
+ * @brief       Board specific definitions for the MIPS Malta FPGA System.
+ *
+ * @author      Imagination Technologies.
```
does it matter, I'm not vain ;-)

haukepetersen commented:
Hmm, I think not setting a default value sounds better to me, but how about some descriptive error message in case the variable is not set?!

haukepetersen commented:
Before I forget, there is one more important problem we have to solve before we can merge this PR: we need to integrate the MIPS toolchain into our CI system. 

@kaspar030: is it easy to integrate the toolchain into Murdock?!

neiljay commented:
>Hmm, I think not setting a default value sounds better to me, but how about some descriptive error 
>message in case the variable is not set?!

done, along with the other changes you requested.

Neill

haukepetersen commented:
Cool, thanks. I already triggered @kaspar030 offline, to include the MIPS toolchain into the CI system, so we can put Murdock onto this PR

kaspar030 commented:
> Cool, thanks. I already triggered @kaspar030 offline, to include the MIPS toolchain into the CI system, so we can put Murdock onto this PR

I'm on int. Are you guys using the installer SDK version, or one of the
binary tarballs?

neiljay commented:
Just the tarballs, you'll want the 'MTI Bare Metal' ones, not 'IMG' ones
(or 'GNU Linux' ones).



On 01/11/17 15:09, Kaspar Schleiser wrote:
>> Cool, thanks. I already triggered @kaspar030 offline, to include the
> MIPS toolchain into the CI system, so we can put Murdock onto this PR
> 
> I'm on int. Are you guys using the installer SDK version, or one of the
> binary tarballs?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/6060#issuecomment-271893051>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AUEU6pmSXVoilK7MfbBXPiXHG4ufN2A2ks5rRPC8gaJpZM4KphNc>.
>

kaspar030 commented:
Thanks @neiljay.

The Docker image update is PR'ed here: https://github.com/RIOT-OS/riotdocker/pull/24

Once that is merged, I'll update the image Murdock is using, and we should be able to test this.

kaspar030 commented:
Murdock should be ready, let's see!

OlegHahm commented:
The CI found a couple of problems.

PeterKietzmann commented:
@neiljay would you please rebase to latest master?

neiljay commented:
The rebase was clean, so I pushed it without a compile check first :-s
But it doesn't build.
It appears at some point, someone has turned on C99 in your build system: -std=c99
This breaks in-line asm code (as this is just a GCC'ism).
I assume you have hit this elsewhere ? What is your preferred solution ?

OlegHahm commented:
See https://github.com/RIOT-OS/RIOT/pull/5824 for the change. `__asm__` should work.

miri64 commented:
> The rebase was clean, so I pushed it without a compile check first :-s

There are a lot of unrelated commits (>250) in this PR now.

haukepetersen commented:
@neiljay would you mind to rebase again? Thx

neiljay commented:
Rebased again, fixed the build issues with the c99 change and the power management API changes.
I see this has now kicked off a CI build, how can I monitor the progress of this - will I get a build log if it fails ?

OlegHahm commented:
> I see this has now kicked off a CI build, how can I monitor the progress of this

AFAIK the Murdock's web frontend currently does not support this. @kaspar030, may correct me if I'm wrong,

>  will I get a build log if it fails ?

Here you go: https://ci.riot-labs.de/RIOT-OS/RIOT/6060/99439487cea2afdbffa1d687cbbf90a978096567/output.html#msp430

However, I see only the error messages for the static tests (trailing white spaces), but not for the various MIPS problems. @kaspar030, @cgundogan?

OlegHahm commented:
Also: https://github.com/RIOT-OS/RIOT/pull/6060/commits/494df5622ec35597146ac82d24f2ad1ede8cd671 seems to create problems for the MSP430 platforms. I would propose to factor this commit out into a separate PR.

cgundogan commented:
you can find the logs by clicking on the `Details` link next to the commit status messages in the bottom section of this page. As for Jenkins (https://riot-ci.inet.haw-hamburg.de/job/RIOT_PR/job/PR-6060/2/artifact/) `write` seems to be undeclared for the `fmt` module on msp boards

OlegHahm commented:
So far, so good, Cenk, but where are the MIPS results? In Murdock I see only some fails, but no error log for these fails and in Jenkins I cannot find the MIPS board at all.

cgundogan commented:
```
15:50:04 [linux_unittests_mips-malta] /opt/jenkins/workspace/RIOT_PR_PR-6060-XRTLXVAEA5TDBZ65VL4HJ3TYDZKODSSMHIEBM22P5TFB5FG4U6YQ@10/cpu/mips32r2_common/Makefile.include:2: *** "Please set $(MIPS_ELF_ROOT) and ensure $(MIPS_ELF_ROOT)/bin is on your PATH".  Stop.
```

Hmm, the path to the toolchain is missing .. will look into that (in Jenkins).

neiljay commented:
Yes no error messages from the mips build ?

I’m confused about this whitespace error message:


“new blank line at EOF”


Haven’t you just enabled std=c99 where a newline at the EOF is a requirement ??

Neil

From: Oleg Hahm [mailto:notifications@github.com]
Sent: 26 January 2017 14:37
To: RIOT-OS/RIOT
Cc: Neil Jones; Mention
Subject: Re: [RIOT-OS/RIOT] mips: Import of mips32r2_common module (#6060)


I see this has now kicked off a CI build, how can I monitor the progress of this

AFAIK the Murdock's web frontend currently does not support this. @kaspar030<https://github.com/kaspar030>, may correct me if I'm wrong,

will I get a build log if it fails ?

Here you go: https://ci.riot-labs.de/RIOT-OS/RIOT/6060/99439487cea2afdbffa1d687cbbf90a978096567/output.html#msp430

However, I see only the error messages for the static tests (trailing white spaces), but not for the various MIPS problems. @kaspar030<https://github.com/kaspar030>, @cgundogan<https://github.com/cgundogan>?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/RIOT-OS/RIOT/pull/6060#issuecomment-275403092>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AUEU6mCyI1Y2WSzU2gQfrThzBdjO8YF-ks5rWK-EgaJpZM4KphNc>.

cgundogan commented:
ok the mips toolchain is installed correctly now. Here you go: https://riot-ci.inet.haw-hamburg.de/job/RIOT_PR/job/PR-6060/6/artifact/

OlegHahm commented:
> Yes no error messages from the mips build ? I’m confused about this whitespace error message: “new blank line at EOF” Haven’t you just enabled std=c99 where a newline at the EOF is a requirement ??

The files that are marked by the CI end with two newlines, e.g.
```Python
In [1]: fd = open("boards/mips-malta/include/periph_conf.h", 'r')

In [2]: fd.read()
Out[2]: '/*\n * Copyright 2016, Imagination Technologies Limited and/or its\n *                 affiliated group companies.\n *\n * This file is subject to the terms and conditions of the GNU Lesser\n * General Public License v2.1. See the file LICENSE in the top level\n * directory for more details.\n *\n */\n\n/**\n * @defgroup    boards_mips-malta MIPS MALTA\n * @ingroup     boards\n * @brief       peripheral configuration for the MIPS Malta FPGA system\n * @{\n *\n * @file\n * @brief       peripheral configuration for the MIPS Malta FPGA system\n *\n * @author      Imagination Technologies.\n */\n\n#ifndef _PERIPH_CONF_H_\n#define _PERIPH_CONF_H_\n\n#ifdef __cplusplus\nextern "C" {\n#endif\n\n/**\n * @brief   Timer definitions\n * @{\n */\n#define TIMER_NUMOF         (1)\n#define TIMER_0_CHANNELS    (3)\n/** @} */\n\n/**\n * @brief No UART driver for this board currently\n *        Note this value must be set though (to 0)\n */\n#define UART_NUMOF          (0)\n\n/**\n * @brief Enable DSP context save + restore.\n */\n#define MIPS_DSP            (1)\n\n/**\n * @brief Enable FPU context save + restore.\n */\n#define MIPS_HARD_FLOAT     (1)\n\n#ifdef __cplusplus\n}\n#endif\n\n#endif /*_PERIPH_CONF_H_*/\n/** @} */\n\n'
```

cgundogan commented:
```
od.c: In function '_bytes_format':
od.c:98:13: error: format '%d' expects argument of type 'int', but argument 3 has type 'long unsigned int' [-Werror=format=]
             sprintf(format, " %%0%do", sizeof(short) * _OCTAL_BYTE_LENGTH);
             ^
od.c:102:13: error: format '%d' expects argument of type 'int', but argument 3 has type 'long unsigned int' [-Werror=format=]
             sprintf(format, " %%0%dlo", sizeof(long) * _OCTAL_BYTE_LENGTH);
             ^
od.c:132:13: error: format '%d' expects argument of type 'int', but argument 3 has type 'long unsigned int' [-Werror=format=]
             sprintf(format, " %%%dd", sizeof(short) * _INT_BYTE_LENGTH);
             ^
od.c:136:13: error: format '%d' expects argument of type 'int', but argument 3 has type 'long unsigned int' [-Werror=format=]
             sprintf(format, " %%%dld", sizeof(long) * _INT_BYTE_LENGTH);
             ^
od.c:170:13: error: format '%u' expects argument of type 'unsigned int', but argument 3 has type 'long unsigned int' [-Werror=format=]
             sprintf(format, " %%%ulu", sizeof(long) * _INT_BYTE_LENGTH);
```

in `sys/od/od.c` for board `mips-malta`

cgundogan commented:
@kaspar030 friendly reminder that murdock and/or the buildtest make target needs some adjustments to include the mips-malta board.

neiljay commented:
Ok nearly there, 3 tests failing, 2 are easy fixes and I'm surprised other arches haven't shown the same issue (just missing deps like stdio.h). But slip requires a UART, what if you don't have one ? can I black-list this test somehow ?

cgundogan commented:
@neiljay you could add `BOARD_BLACKLIST += mips-malta` to the `Makefile` of `tests/slip` like it's e.g. done in `tests/shell/Makefile`

neiljay commented:
@cgundogan 

It still try's to build with this, just issuing a warning (i also tried FEATURES_REQUIRED):

```
$ make
The selected BOARD=mips-malta is blacklisted: mips-malta


EXPECT ERRORS!


Building application "tests_driver_slip" for "mips-malta" with MCU "mips32r2_common".

"/usr/bin/make" -C /user/neil.jones/src/RIOT/boards/mips-malta
"/usr/bin/make" -C /user/neil.jones/src/RIOT/core
"/usr/bin/make" -C /user/neil.jones/src/RIOT/cpu/mips32r2_common
"/usr/bin/make" -C /user/neil.jones/src/RIOT/cpu/mips32r2_common/periph
"/usr/bin/make" -C /user/neil.jones/src/RIOT/drivers
"/usr/bin/make" -C /user/neil.jones/src/RIOT/sys
"/usr/bin/make" -C /user/neil.jones/src/RIOT/sys/auto_init
"/usr/bin/make" -C /user/neil.jones/src/RIOT/sys/auto_init/netif
"/usr/bin/make" -C /user/neil.jones/src/RIOT/sys/net/gnrc
"/usr/bin/make" -C /user/neil.jones/src/RIOT/sys/net/gnrc/link_layer/slip
"/usr/bin/make" -C /user/neil.jones/src/RIOT/sys/net/gnrc/netapi
"/usr/bin/make" -C /user/neil.jones/src/RIOT/sys/net/gnrc/netif
"/usr/bin/make" -C /user/neil.jones/src/RIOT/sys/net/gnrc/netif/hdr
"/usr/bin/make" -C /user/neil.jones/src/RIOT/sys/net/gnrc/netreg
"/usr/bin/make" -C /user/neil.jones/src/RIOT/sys/net/gnrc/pkt
"/usr/bin/make" -C /user/neil.jones/src/RIOT/sys/net/gnrc/pktbuf_static
"/usr/bin/make" -C /user/neil.jones/src/RIOT/sys/net/gnrc/pktdump
"/usr/bin/make" -C /user/neil.jones/src/RIOT/sys/od
"/usr/bin/make" -C /user/neil.jones/src/RIOT/sys/shell
"/usr/bin/make" -C /user/neil.jones/src/RIOT/sys/shell/commands
/user/neil.jones/src/RIOT/tests/slip/bin/mips-malta/gnrc_slip.a(gnrc_slip.o): In function `_slip_send_char':
/user/neil.jones/src/RIOT/sys/net/gnrc/link_layer/slip/gnrc_slip.c:(.text._slip_send_char+0x14): undefined reference to `uart_write'
/user/neil.jones/src/RIOT/tests/slip/bin/mips-malta/gnrc_slip.a(gnrc_slip.o): In function `gnrc_slip_init':
/user/neil.jones/src/RIOT/sys/net/gnrc/link_layer/slip/gnrc_slip.c:(.text.gnrc_slip_init+0x58): undefined reference to `uart_init'
collect2: error: ld returned 1 exit status
make: *** [all] Error 1
```
Will the CI builds honour this ?

OlegHahm commented:
Yes, this CI calls `make buildtest` which excludes blacklisted boards.

PeterKietzmann commented:
@neiljay I will move the the milestone label now as we will 'feature freeze' in about the next minutes. That must not mean that your PR can't be merged to master once approved and built correctly. Thanks for this great feature btw :-)

cgundogan commented:
>Yes, this CI calls make buildtest which excludes blacklisted boards.

True for Murdock (: In Jenkins however the build is only triggered, if `make info-boards-supported` returns the board in question. `make info-boards-supported` should honour the `BLACKLIST_BOARD` variable.

neiljay commented:
@OlegHahm 
Whats your workflow wrt sqashing? will you do that prior to the final merge into a single commit ?

OlegHahm commented:
As stated in the [Development Procedures](https://github.com/RIOT-OS/RIOT/wiki/Development-procedures) you should wait with squashing until a maintainer approves the PR. You don't have to squash everything into one commit. For instance, I would definitely keep the fixes outside `cpu` and `board` as separate commits.

neiljay commented:
Hi,

We are now failing because of this:

/user/neil.jones/src/RIOT/sys/shell/commands/sc_gnrc_6ctx.c:88:5: error: implicit declaration of function 'strtok_r' [-Werror=implicit-function-declaration]
     addr_str = strtok_r(prefix_str, "/", &save_ptr);

strtok_r is not c99:

http://stackoverflow.com/questions/23961147/implicit-declaration-of-function-strtok-r-wimplicit-function-declaration-in

@miri64 looks like this is your file,  I think you need to set #define _POSIX_C_SOURCE appropriately not _DEFAULT_SOURCE, to be more portable. Or just use std=gnu99 and most of these c99 issues go away.  In fact the only way I could get this to work with our C-lib is to '#undef __STRICT_ANSI__'

This test needs a UART anyway (which we don't have) so I'll blacklist it anyway.

kaspar030 commented:
> Or just use std=gnu99 and most of these c99 issues go away.

You can set std=gnu99 just for mips, if necessary.

neiljay commented:
Ok so Jenkins is passing at last, Murdock is complaining that it has a 'Need Squashing'  label which I'm happy to do once approved. Not sure why the mips32r3 build is failing - no error messages and its passing in Jenkins ??

Can we finally get this merged before the world changes again.

neiljay commented:
@kaspar030 Any help with why murdock is failing ?

OlegHahm commented:
@miri64, only saw this now: any particular reason for the "major" label?

neiljay commented:
@miri64 @kaspar030 @OlegHahm Hello ;-) ;-)

miri64 commented:
@OlegHahm well it is a pretty major new feature isn't it?

kaspar030 commented:
Something's wrong with the buildtest...

kaspar030 commented:
> Something's wrong with the buildtest...

Could be the mips env variables, are they passed through in buildtest?

OlegHahm commented:
> @OlegHahm well it is a pretty major new feature isn't it?

Well, that's not exactly the point of the "major" label. It says that this PR requires a very thorough review and more than one maintainer ACK.

cgundogan commented:
@kaspar030 can you check whether the user that executes the build in Docker has the right env variables? In the [Dockerfile](https://github.com/RIOT-OS/riotdocker/blob/master/Dockerfile) the `PATH` is only adjusted for the root user.
```
ENV PATH $PATH:/opt/mips-mti-elf/2016.05-03/bin
ENV MIPS_ELF_ROOT /opt/mips-mti-elf/2016.05-03
```
If murdock uses another user in the docker, then the `PATH` won't contain these values.

kaspar030 commented:
@neiljay please try https://github.com/neiljay/RIOT/pull/3.

kaspar030 commented:
@cgundogan Looks ok:

```
[kaspar@booze hello-world (pr6060)]$ docker run -ti -u 1000 riot/riotbuild:latest /bin/bash
riotbuild@1fae0df0da87:~$ env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt/mips-mti-elf/2016.05-03/bin
MIPS_ELF_ROOT=/opt/mips-mti-elf/2016.05-03
riotbuild@1fae0df0da87:~$
```

(I cut out some)

neiljay commented:
Cool its now passing in Murdock, the failure is due to the "Need Squashing" Label.

Are we all happy now ? I'll update the author to me then squash ?

Cheers,

Neil

kaspar030 commented on tests/leds/main.c:
```diff
@@ -19,10 +19,12 @@
 #include "board.h"
 #include "periph_conf.h"
 
+
```
this one wasn't here before...

neiljay replied:
hmm how did that get in, I'll fix.

neiljay commented:
@haukepetersen  Can you please revert your review request so we can finally get this merged!

> 
haukepetersen on Dec 12, 2016 Member

these addresses are CPU dependent, right? So they should go into the CPU code?!
@neiljay
neiljay on Dec 12, 2016 Member

No, they are specific to the Malta FPGA platform only.
@haukepetersen
haukepetersen on Dec 13, 2016 Member

ah ok.

gebart commented:
Please update `dist/tools/ci/print_toolchain_versions.sh` to include a version line for the compiler and libc. This will let you see the toolchain version in the output from the static tests on Jenkins and Murdock, makes it easier to track down problems.

neiljay commented:
@gebart In a separate PR ?

gebart commented:
@neiljay I'm fine with having the info script update in a separate PR.

neiljay commented:
@gebart see  #6585

kaspar030 commented:
@haukepetersen I'll dismiss your review, reading your comment it seems your comment has been addressed.

kaspar030 commented:
Congrats @neiljay for adding the first new architecture in years! Thanks for all the work!

neiljay commented:
Thanks all !

francois-berder-imgtec commented:
Thank you very much everyone for your great work !

miri64 commented:
At last! Congratz and thanks for your big contribution and your patience @neiljay :party: :sparkles:

emmanuelsearch commented:
yay! Congrats!

kYc0o commented:
Wow congrats! I'm looking forward to test it! Thanks to all people here.

gebart commented:
Congratulations on getting this large PR merged!
@neiljay I did run into some trouble trying to build this on my machine, as well as when testing #5616 on Jenkins. I have create a tracking issue for the issues in #6588. Please don't take this as a lack of appreciation, I think it's great that we are getting new hardware supported, I only want to improve on it.
